File: SearchMD.py
import undetected_chromedriver as uc
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException
import time
import random
import os
import sys # Import sys to check the operating system
import json
import subprocess
import vosk
import requests
import shutil

# --- Configuration (kept for CAPTCHA routine) ---
VOSK_MODEL_PATH = os.path.join(os.path.dirname(__file__), 'vosk-model-small-en-us-0.15')
DOWNLOAD_PATH = os.path.join(os.path.dirname(__file__), 'downloads')
AUDIO_MP3_PATH = os.path.join(DOWNLOAD_PATH, 'captcha_audio_md.mp3')
SAMPLE_RATE = 16000

# --- Helper Functions ---
def humanlike_type(driver, selector, text):
    element = driver.find_element(*selector)
    for char in text:
        element.send_keys(char)
        time.sleep(random.uniform(0.05, 0.12))

def transcribe_audio(file_path):
    wav_file_path = os.path.join(os.path.dirname(file_path), 'audio_md.wav')
    ffmpeg_command = f'ffmpeg -i "{file_path}" -ar {SAMPLE_RATE} -ac 1 "{wav_file_path}" -y'
    subprocess.run(ffmpeg_command, shell=True, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    model = vosk.Model(VOSK_MODEL_PATH)
    rec = vosk.KaldiRecognizer(model, SAMPLE_RATE)
    with open(wav_file_path, "rb") as wf:
        while True:
            data = wf.read(4000)
            if len(data) == 0: break
            rec.AcceptWaveform(data)
    result = json.loads(rec.FinalResult())
    os.remove(wav_file_path)
    return result['text']

def check_maryland_dependencies():
    """Checks for dependencies required by the CAPTCHA routine."""
    if not os.path.exists(VOSK_MODEL_PATH):
        return { "error": "Dependency Missing: Vosk Model", "details": "The folder 'vosk-model-small-en-us-0.15' was not found." }
    if not shutil.which('ffmpeg'):
        return { "error": "Dependency Missing: FFmpeg", "details": "FFmpeg could not be found in your system's PATH." }
    return None

# --- *** NEW HELPER FUNCTION TO FIND CHROME *** ---
def get_chrome_executable_path():
    """Finds the default path for Google Chrome to help undetected_chromedriver."""
    if sys.platform == 'win32': # For Windows
        for path in [
            os.path.join(os.environ["ProgramFiles(x86)"], "Google", "Chrome", "Application", "chrome.exe"),
            os.path.join(os.environ["ProgramFiles"], "Google", "Chrome", "Application", "chrome.exe"),
        ]:
            if os.path.exists(path):
                return path
    elif sys.platform == 'linux': # For Linux (including WSL)
        for path in ["/usr/bin/google-chrome-stable", "/usr/bin/google-chrome"]:
             if os.path.exists(path):
                return path
    # Add macOS path if needed
    # elif sys.platform == 'darwin':
    #     path = "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"
    #     if os.path.exists(path):
    #         return path
    return None

# --- Main Scraper Function ---
def search_md(search_args):
    dependency_error = check_maryland_dependencies()
    if dependency_error:
        return dependency_error

    entity_name_to_search = search_args.get("entity_name")
    if not entity_name_to_search:
        return {"error": "Entity name is required for Maryland search."}

    os.makedirs(DOWNLOAD_PATH, exist_ok=True)
    driver = None
    
    try:
        chrome_path = get_chrome_executable_path()
        if not chrome_path:
            return {"error": "Could not find Google Chrome executable. Please ensure it is installed."}

        # --- THIS IS THE FIX ---
        # We now explicitly provide the path to the Chrome browser.
        driver = uc.Chrome(browser_executable_path=chrome_path)
        # --- END OF FIX ---
        
        wait = WebDriverWait(driver, 45)

        driver.get('https://egov.maryland.gov/BusinessExpress/EntitySearch')

        humanlike_type(driver, (By.ID, 'BusinessName'), entity_name_to_search)
        driver.find_element(By.ID, 'searchBus1').click()

        results_table_selector = (By.ID, 'newTblBusSearch')
        try:
            WebDriverWait(driver, 15).until(EC.visibility_of_element_located(results_table_selector))
        except TimeoutException:
            try:
                # ... (rest of the CAPTCHA logic remains the same) ...
                try:
                    short_wait = WebDriverWait(driver, 5)
                    anchor_frame = short_wait.until(EC.presence_of_element_located((By.CSS_SELECTOR, 'iframe[src*="api2/anchor"]')))
                    driver.switch_to.frame(anchor_frame)
                    wait.until(EC.element_to_be_clickable((By.ID, 'recaptcha-anchor'))).click()
                    driver.switch_to.default_content()
                except TimeoutException:
                    driver.switch_to.default_content()
                
                bframe_handle = wait.until(EC.presence_of_element_located((By.CSS_SELECTOR, 'iframe[src*="api2/bframe"]')))
                driver.switch_to.frame(bframe_handle)
                audio_button_element = wait.until(EC.element_to_be_clickable((By.ID, 'recaptcha-audio-button')))
                driver.execute_script("arguments[0].click();", audio_button_element)
                time.sleep(3) 
                driver.switch_to.default_content()
                driver.switch_to.frame(bframe_handle)
                
                audio_url = wait.until(EC.presence_of_element_located((By.CSS_SELECTOR, '.rc-audiochallenge-tdownload-link'))).get_attribute('href')
                response = requests.get(audio_url, stream=True)
                with open(AUDIO_MP3_PATH, 'wb') as f:
                    for chunk in response.iter_content(chunk_size=1024):
                        if chunk: f.write(chunk)
                solution_text = transcribe_audio(AUDIO_MP3_PATH)
                if not solution_text: raise Exception("Transcription failed.")
                
                driver.find_element(By.ID, 'audio-response').send_keys(solution_text)
                wait.until(EC.element_to_be_clickable((By.ID, 'recaptcha-verify-button'))).click()
                driver.switch_to.default_content()
                wait.until(EC.visibility_of_element_located(results_table_selector))
            except Exception as e:
                raise Exception(f"CAPTCHA solver failed: {e}")

        original_window = driver.current_window_handle
        first_result_selector = (By.CSS_SELECTOR, '#newTblBusSearch > tbody > tr:nth-child(1) > td:nth-child(2) > a')
        wait.until(EC.element_to_be_clickable(first_result_selector)).click()
        
        wait.until(EC.number_of_windows_to_be(2))
        for window_handle in driver.window_handles:
            if window_handle != original_window:
                driver.switch_to.window(window_handle)
                break
        
        wait.until(EC.visibility_of_element_located((By.CSS_SELECTOR, 'div.fp_formItemGroup')))
        
        js_get_value_by_label = """
            const label = arguments[0];
            const allLabels = document.querySelectorAll('div.fp_formItem strong');
            for (const strongEl of allLabels) {
                if (strongEl.innerText.trim().replace(':', '') === label) {
                    let valueEl = strongEl.parentElement.nextElementSibling;
                    if (valueEl && (valueEl.classList.contains('fp_formItemData') || valueEl.querySelector('.fp_formItemData'))) {
                         valueEl = valueEl.querySelector('.fp_formItemData') || valueEl;
                    }
                    if (valueEl) {
                        return valueEl.innerHTML.replace(/<br\\s*[/]?>/gi, ' ').replace(/\\s+/g, ' ').trim();
                    }
                }
            }
            return null;
        """

        def get_detail(label):
            return driver.execute_script(js_get_value_by_label, label)

        entity_name_b = get_detail("Business Name")
        owner_details = get_detail("Owner")
        address_po = get_detail("Principal Office")
        entity_status = get_detail("Status")
        entity_name, address = None, None

        if entity_name_b:
            entity_name = entity_name_b
        elif owner_details:
            parts = owner_details.split(); name_parts = []; address_start_index = -1
            for i, part in enumerate(parts):
                if part.isdigit() and len(part) > 2:
                    address_start_index = i; break
                name_parts.append(part)
            if address_start_index != -1:
                entity_name = ' '.join(name_parts); address = ' '.join(parts[address_start_index:])
            else:
                entity_name = owner_details
        if address_po:
            address = address_po

        scraped_data = {
            "entity_name": entity_name,
            "registration_date": get_detail("Date of Formation/ Registration"),
            "entity_type": get_detail("Business Type"),
            "business_identification_number": get_detail("Department ID Number"),
            "entity_status": entity_status,
            "statusActive": entity_status and 'active' in entity_status.lower(),
            "address": address or ""
        }
        
        driver.close()
        driver.switch_to.window(original_window)
        
        return [scraped_data]
            
    except Exception as e:
        error_dir = os.path.join(os.path.dirname(__file__), "errors")
        os.makedirs(error_dir, exist_ok=True)
        screenshot_path = os.path.join(error_dir, f"maryland_unexpected_error_{int(time.time())}.png")
        if driver: driver.save_screenshot(screenshot_path)
        return {"error": "An unexpected error occurred.", "details": str(e)}
            
    finally:
        if driver:
            driver.quit()

File: SearchMD.js
// --- Dependencies ---
const puppeteer = require('puppeteer-extra');
const StealthPlugin = require('puppeteer-extra-plugin-stealth');
puppeteer.use(StealthPlugin());

const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');
const vosk = require('vosk');
const axios = require('axios');

// --- Configuration ---
const VOSK_MODEL_PATH = 'vosk-model-small-en-us-0.15'; 
const DOWNLOAD_PATH = path.resolve(__dirname, 'downloads');
const AUDIO_MP3_PATH = path.join(DOWNLOAD_PATH, 'captcha_audio_md.mp3');
const SAMPLE_RATE = 16000;
const ERROR_PATH = path.resolve(__dirname, 'errors');

// --- Helper Functions ---
const randomDelay = (min = 500, max = 1200) => new Promise(resolve => setTimeout(resolve, Math.random() * (max - min) + min));

const transcribeAudio = (filePath) => {
    return new Promise((resolve, reject) => {
        const wavFilePath = path.join(path.dirname(filePath), 'audio_md.wav');
        const ffmpegCommand = `ffmpeg -i "${filePath}" -ar ${SAMPLE_RATE} -ac 1 "${wavFilePath}" -y`;
        exec(ffmpegCommand, (error) => {
            if (error) return reject(`FFmpeg error: ${error}`);
            const model = new vosk.Model(VOSK_MODEL_PATH);
            const rec = new vosk.Recognizer({ model: model, sampleRate: SAMPLE_RATE });
            const stream = fs.createReadStream(wavFilePath);
            stream.on('data', (chunk) => rec.acceptWaveform(chunk));
            stream.on('end', () => {
                const result = rec.finalResult();
                rec.free(); model.free();
                fs.unlinkSync(wavFilePath);
                resolve(result.text.trim());
            });
            stream.on('error', (err) => reject(`Error reading audio stream: ${err}`));
        });
    });
};


// --- Main Automation Logic ---
const scrapeMaryland = async (searchTerm, outputFilename) => {
    if (!fs.existsSync(VOSK_MODEL_PATH)) {
        console.error(`[JS ERROR] Vosk model not found at path: ${VOSK_MODEL_PATH}`);
        fs.writeFileSync(outputFilename, JSON.stringify([]));
        return;
    }
    if (!fs.existsSync(DOWNLOAD_PATH)) fs.mkdirSync(DOWNLOAD_PATH);
    if (!fs.existsSync(ERROR_PATH)) fs.mkdirSync(ERROR_PATH);

    let browser;
    // --- FIX: Declare page outside the try block for error handling ---
    let page;
    try {
        browser = await puppeteer.launch({
            headless: false,
            args: [
                '--no-sandbox', '--disable-setuid-sandbox', '--disable-infobars', '--disable-dev-shm-usage',
                '--disable-blink-features=AutomationControlled', '--start-maximized',
            ],
            ignoreDefaultArgs: ['--enable-automation']
        });

        page = await browser.newPage();
        page.setDefaultTimeout(60000);
        await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36');
        await page.setViewport({ width: 1920, height: 1080 });

        console.log("Navigating to the Business Entity Search page...");
        await page.goto('https://egov.maryland.gov/BusinessExpress/EntitySearch', { waitUntil: 'networkidle2' });
        await randomDelay();

        console.log(`Typing "${searchTerm}" into the search field...`);
        await page.type('#BusinessName', searchTerm, { delay: 100 });
        await randomDelay(300, 700);
        
        console.log("Clicking search...");
        await page.click('#searchBus1');

        console.log("Watching for page outcome (CAPTCHA or results)...");
        const captchaIframeSelector = 'iframe[title="reCAPTCHA"]';
        const resultsTableSelector = '#newTblBusSearch';
        const noResultsSelector = 'div.p-3';

        await page.waitForSelector(`${captchaIframeSelector}, ${resultsTableSelector}, ${noResultsSelector}`);

        if (await page.$(noResultsSelector)) {
            const noResultsText = await page.$eval(noResultsSelector, el => el.innerText);
            if (noResultsText.includes('No business filings were found')) {
                console.log("Search returned no results. Script finishing successfully.");
                fs.writeFileSync(outputFilename, JSON.stringify([]));
                await browser.close();
                return;
            }
        }
        
        if (await page.$(captchaIframeSelector)) {
            console.log("CAPTCHA detected. Starting solver...");
            
            // --- *** THE FIX IS HERE *** ---
            // Step 1: Get the iframe handle and its content frame.
            const anchorFrameHandle = await page.waitForSelector('iframe[src*="api2/anchor"]');
            const anchorFrame = await anchorFrameHandle.contentFrame();

            // Step 2: Explicitly wait for the checkbox to be visible *within that frame*.
            console.log("Waiting for CAPTCHA checkbox to render inside its iframe...");
            await anchorFrame.waitForSelector('#recaptcha-anchor', { visible: true, timeout: 10000 });
            
            // Step 3: Use the frame's own click method. This is the crucial part.
            console.log("Clicking CAPTCHA checkbox...");
            await anchorFrame.click('#recaptcha-anchor');
            await randomDelay(2000, 3000); // Wait for challenge to appear
            // --- *** END OF FIX *** ---

            try {
                const bframeHandle = await page.waitForSelector('iframe[src*="api2/bframe"]', { visible: true, timeout: 5000 });
                const bframe = await bframeHandle.contentFrame();
                
                await bframe.waitForSelector('#recaptcha-audio-button', { visible: true });
                await bframe.click('#recaptcha-audio-button');
                await randomDelay();

                const audioLinkSelector = '.rc-audiochallenge-tdownload-link';
                await bframe.waitForSelector(audioLinkSelector, { visible: true });
                
                const audioUrl = await bframe.$eval(audioLinkSelector, el => el.href);
                const response = await axios.get(audioUrl, { responseType: 'stream' });
                const writer = fs.createWriteStream(AUDIO_MP3_PATH);
                response.data.pipe(writer);
                await new Promise((resolve, reject) => { writer.on('finish', resolve); writer.on('error', reject); });
                
                const solutionText = await transcribeAudio(AUDIO_MP3_PATH);
                if (!solutionText) throw new Error("Transcription failed.");
                console.log(`Transcription result: "${solutionText}"`);
                
                await bframe.type('#audio-response', solutionText, { delay: 100 });
                await randomDelay();
                
                await bframe.waitForSelector('#recaptcha-verify-button', { visible: true });
                await bframe.click('#recaptcha-verify-button');

                console.log("✅ CAPTCHA solved. Waiting for results...");
            } catch (e) {
                console.log("Instant verification or no challenge needed. Proceeding...");
            }
        } else {
            console.log("No CAPTCHA detected. Proceeding directly to results.");
        }
        
        await page.waitForSelector(resultsTableSelector, { visible: true, timeout: 30000 });
        console.log("Search results page loaded.");
        
        await randomDelay();
        const firstResultSelector = '#newTblBusSearch > tbody > tr:nth-child(1) > td:nth-child(2) > a';
        
        await page.waitForSelector(firstResultSelector, { visible: true, timeout: 10000 });
        const newPagePromise = browser.waitForTarget(target => target.opener() === page.target()).then(target => target.page());
        await page.click(firstResultSelector);
        const newPage = await newPagePromise;
        if (!newPage) { throw new Error("Could not find the new page/tab."); }
        
        await newPage.setDefaultTimeout(60000);
        await newPage.setViewport({ width: 1920, height: 1080 });
        
        const detailsContainerSelector = '.fp_formItemGroup';
        await newPage.waitForSelector(`${detailsContainerSelector} .fp_formItemLabel strong`, { visible: true });

        const businessData = await newPage.evaluate((containerSelector) => {
            const getTextByLabel = (label) => { const allLabels = Array.from(document.querySelectorAll(`${containerSelector} .fp_formItemLabel strong`)); const targetLabel = allLabels.find(l => l.textContent.trim() === label); if (targetLabel) { const dataElement = targetLabel.closest('.fp_formItem').querySelector('.fp_formItemData'); return dataElement ? dataElement.innerText.trim().replace(/\s\s+/g, ' ') : null; } return null; };
            const entity_status = getTextByLabel('Status:');
            const activeKeywords = ['ACTIVE', 'IN GOOD STANDING', 'INCORPORATED'];
            const statusActive = entity_status ? activeKeywords.some(keyword => entity_status.toUpperCase().includes(keyword)) : false;
            return {
                entity_name: getTextByLabel('Business Name:'),
                registration_date: getTextByLabel('Date of Formation/ Registration:'),
                entity_type: getTextByLabel('Business Type:'),
                business_identification_number: getTextByLabel('Department ID Number:'),
                entity_status: entity_status,
                statusActive: statusActive,
                address: getTextByLabel('Principal Office:')
            };
        }, detailsContainerSelector);

        await newPage.close();
        
        console.log("Scraping complete. Writing data to file.");
        fs.writeFileSync(outputFilename, JSON.stringify([businessData], null, 2));

    } catch (err) {
        const timestamp = new Date().getTime();
        const screenshotPath = path.join(ERROR_PATH, `error-maryland-${timestamp}.png`);
        console.error(`An error occurred: ${err.message}\n${err.stack}`);
        console.log(`Saving screenshot to ${screenshotPath}`);
        
        if(page && !page.isClosed()) {
            await page.screenshot({ path: screenshotPath, fullPage: true });
        }
        
        fs.writeFileSync(outputFilename, JSON.stringify([]));

    } finally {
        if (browser) {
            await browser.close();
        }
    }
};

// --- Script Execution ---
const searchTerm = process.argv[2];
const outputFilename = process.argv[3];

if (!searchTerm || !outputFilename) {
    console.error("Error: Missing searchTerm or outputFilename arguments.");
    if(outputFilename) fs.writeFileSync(outputFilename, JSON.stringify([]));
    process.exit(1);
}
console.log(`Starting Maryland scrape for term: "${searchTerm}"`);
scrapeMaryland(searchTerm, outputFilename);